internal/generate/crd/crd.go

```diff
@@ -0,0 +1,332 @@
+func NewCRDNonGo(cfg genutil.Config, resource scaffold.Resource) genutil.Generator {
+	g := crdGenerator{
+		Config:       cfg,
+		resource:     resource,
+		isOperatorGo: false,
+	}
+	if g.InputDir == "" {
+		g.InputDir = scaffold.CRDsDir
+	}
+	if g.OutputDir == "" {
+		g.OutputDir = scaffold.CRDsDir
+	}
+	return g
+}
+
+func (g crdGenerator) validate() error {
+	if g.InputDir == "" {
+		return errors.New("input dir cannot be empty")
+	}
+	if g.InputDir == "" {
+		return errors.New("output dir cannot be empty")
+	}
+	if !g.isOperatorGo {
+		if err := g.resource.Validate(); err != nil {
+			return errors.Wrap(err, "resource is invalid:")
+		}
+	}
+	return nil
+}
+
+// Generate generates CRD manifests and writes them to g.OutputDir.
+func (g crdGenerator) Generate() (err error) {
+	if err = g.validate(); err != nil {
+		return errors.Wrap(err, "validation error")
+	}
+	var fileMap map[string][]byte
+	if g.isOperatorGo {
+		fileMap, err = g.generateGo()
+	} else {
+		fileMap, err = g.generateNonGo()
+	}
+	if err != nil {
+		return errors.Wrap(err, "error generating CRD manifests")
+	}
+	if err = os.MkdirAll(g.OutputDir, fileutil.DefaultDirFileMode); err != nil {
+		return errors.Wrapf(err, "error mkdir %s", g.OutputDir)
+	}
+	for fileName, b := range fileMap {
+		path := filepath.Join(g.OutputDir, fileName)
+		if err := ioutil.WriteFile(path, b, fileutil.DefaultFileMode); err != nil {
+			return errors.Wrap(err, "error writing CRD manifests")
+		}
+	}
+	return nil
+}
+
+func getFileNameForResource(r scaffold.Resource) string {
+	return fmt.Sprintf("%s_%s_crd.yaml", r.FullGroup, r.Resource)
+}
+
+// generateNonGo generates CRDs for Go projects using Go API files.
+func (g crdGenerator) generateGo() (map[string][]byte, error) {
+	fileMap := map[string][]byte{}
+	// Generate files in the generator's cache so we can modify the file name
+	// and annotations.
+	defName := "output:crd:cache"
+	cacheOutputDir := string(filepath.Separator) + filepath.Clean(g.OutputDir)
+	rawOpts := []string{
+		"crd",
+		fmt.Sprintf("paths=%s/...", fileutil.DotPath(g.InputDir)),
+		fmt.Sprintf("%s:dir=%s", defName, cacheOutputDir),
+	}
+	cachedGen := genutil.NewCachedGenerator()
+	cachedGen.AddOutputRule(defName, genutil.OutputToCachedDirectory{})
+	if err := cachedGen.Run(rawOpts); err != nil {
+		return nil, err
+	}
+	cache := genutil.GetCache()
+	infos, err := afero.ReadDir(cache, cacheOutputDir)
+	if err != nil {
+		return nil, err
+	}
+	for _, info := range infos {
+		if info.IsDir() {
+			continue
+		}
+		path := filepath.Join(cacheOutputDir, info.Name())
+		b, err := afero.ReadFile(cache, path)
+		if err != nil {
+			return nil, err
+		}
+		scanner := yamlutil.NewYAMLScanner(b)
+		modifiedCRD := []byte{}
+		for scanner.Scan() {
+			sb := scanner.Bytes()
+			typeMeta, err := k8sutil.GetTypeMetaFromBytes(sb)
+			if err != nil {
+				return nil, err
+			}
+			if typeMeta.Kind != "CustomResourceDefinition" {
+				continue
+			}
```

Comment:
Does `controller-gen crd` generate other kinds? Or do we check the kind so we can unmarshal into the CRD type and remove the annotation?

Wondering if it's safer for future-proofing if we don't assume the kind and generically remove the annotation using `unstructured.Unstructured`?

Reply:
The latter. I agree using `unstructured.Unstructured` is a better general solution.